rfc/006-backfill-blocks.md

```diff
@@ -0,0 +1,187 @@
+retrieve and verify blocks up to (and including) the specified backfill height
+**before** participating in consensus.
+
+### Backfill on application request
+
+The application has control of block retention via `retain_height`, called here:
+
+message ResponseCommit {
+  // reserve 1
+  bytes data          = 2;
+  int64 retain_height = 3;
+}
+
+Beforehand, a retain height that was less than the nodes current base was
+ignored. With the same backfill mechanism, the application should now be able to
+lower the retention height below the current base height (whether to accommodate
+changing consensus parameters, app specific needs, or a local config).
+This accommodates the use case of a truncated node that wishes to become an
+archive node. Note this could be done simultaneously alongside consensus.
+
+### Verification
+
+Nodes will need to verify these prior blocks. This can be achieved by first
+retrieving the header at the base height from the block store. The node then
+checks that the `LastBlockID` corresponds with the hash calculated from the
+header in the new block directly below:
+
+```go
+header[height].LastBlockID == hash(header[height-1])
+```
+
+The node can therefore trust the new header. This then can be followed with
+using the hashes in the trusted header to validate the other block and state
+info. This is recursively done till the backfill height.  
+
+### Design
+
+The backfill mechanism would reside in the blockchain reactor. One may view it
+similarly to fast syncs position within the reactor as a service that can be
+started and stopped. Backfill, however, would not be an optional configuration.
+
+Two new messages would be added to allow for the passing of state data:
+
+```proto
+message StateDataRequest {
+  int64 height = 1;
+}
+```
+
+```proto
+message StateDataResponse {
+  int64 height = 1;
+  ValidatorSet validator_set = 2;
+  ConsensusParams consensus_params = 3;
+  ABCIResponses abci_responses = 4;
+}
+```
+
+This would be sent across the existing blockchain channel. The advantage with
+this setup is that all backfill data is within the same domain, making it
+easier to coordinate both verification of block and state data.
+
+## Alternative Solutions
+
+If we do not want to extend the blockchain's functionality, it is possible to
+use the embedded light client within state sync instead.
+
+This approach would separate the two forms of data. The state sync reactor would
+be responsible for verifying state data and persisting it whilst the blockchain
+reactor would process only the blocks. Operating asynchronously could cause
+inconsistencies in which data was available (if blockchain reactor is faster
+we have blocks with no validator set; if state sync reactor is faster then we
+have validator sets with no blocks). This would also require duplication in
+headers sent (one for the light client and one within the block sent to the
+blockchain reactor).
```

Comment:
agree it would be consensus or at least the block executor passing a call to state sync to retrieve validator sets and consensus params. Sounds a bit messy